parley/src/layout/mod.rs

```diff
@@ -261,3 +267,149 @@ pub struct Decoration<B: Brush> {
+                    node.set_previous_on_line(last_id);
+                    update.nodes.push((last_id, last_node));
+                    parent_node.push_child(last_id);
+                }
+
+                node.set_bounds(accesskit::Rect {
+                    x0: x_offset + run_offset as f64,
+                    y0: y_offset + metrics.min_coord as f64,
+                    x1: x_offset + (run_offset + run.advance()) as f64,
+                    y1: y_offset + metrics.max_coord as f64,
+                });
+                node.set_text_direction(if run.is_rtl() {
+                    TextDirection::RightToLeft
+                } else {
+                    TextDirection::LeftToRight
+                });
+
+                let run_text = &text[run.text_range()];
+                node.set_value(run_text);
+
+                let mut character_lengths = Vec::new();
+                let mut cluster_offset = 0.0;
+                let mut character_positions = Vec::new();
+                let mut character_widths = Vec::new();
+                let mut word_lengths = Vec::new();
+                let mut last_word_start = 0;
+
+                for cluster in run.clusters() {
+                    let cluster_text = &text[cluster.text_range()];
+                    if cluster.is_word_boundary()
+                        && !cluster.is_space_or_nbsp()
+                        && !character_lengths.is_empty()
+                    {
+                        word_lengths.push((character_lengths.len() - last_word_start) as _);
+                        last_word_start = character_lengths.len();
+                    }
+                    character_lengths.push(cluster_text.len() as _);
+                    character_positions.push(cluster_offset);
+                    character_widths.push(cluster.advance());
+                    cluster_offset += cluster.advance();
+                }
+
+                word_lengths.push((character_lengths.len() - last_word_start) as _);
+                node.set_character_lengths(character_lengths);
+                node.set_character_positions(character_positions);
+                node.set_character_widths(character_widths);
+                node.set_word_lengths(word_lengths);
+
+                last_node = Some((id, node));
+            }
+
+            if let Some((id, node)) = last_node {
+                update.nodes.push((id, node));
+                parent_node.push_child(id);
+            }
+        }
+
+        // Remove mappings for runs that no longer exist.
+        let mut ids_to_remove = Vec::<NodeId>::new();
+        let mut run_paths_to_remove = Vec::<(usize, usize)>::new();
+        for (access_id, run_path) in self.run_paths_by_access_id.iter() {
+            if !ids.contains(access_id) {
+                ids_to_remove.push(*access_id);
+                run_paths_to_remove.push(*run_path);
+            }
+        }
+        for id in ids_to_remove {
+            self.run_paths_by_access_id.remove(&id);
```

Comment:
Presumably it's a choice for simplicity to not reuse these ids?

Reply:
Hmm. I guess that makes sense as a pragmatic choice. It's not like there's any shortage of 64 bit ids.

My concern is also that if you e.g. have:
```text
{a bit of text}[0]
{with some importance}[1]
```
(where `{}[id]` marks a run with id `id`)
and then add a newline before the word "text", you end up with:
```text
{a bit of}[0]
{text}[1]
{with some importance}[2]
```
That is, the second run has been completely transformed in content.

I really don't have a good suggestion here, unfortunately. I imagine your best chance of doing better is with something like the editor abstraction, so you know what has changed. But that's an awful lot of bookkeeping work needed.

Reply:
You're completely right about the effects of inserting text. I'm not sure how much this matters. Of the platform backends implemented so far, this only makes a difference at all on Windows, where UIA clients can have long-lived references to text range objects, and in our implementation, those objects point to specific run IDs. I don't know if any UIA implementation really handles this ideally.

Reply:
It's also worth noting that the egui implementation, which has been my de-facto reference implementation until now, has the same problem with the insertion scenario. So the egui implementation is really no better. It uses line indices as the basis of run node IDs. It just doesn't have an explicit mapping, since it generates AccessKit node IDs via hashing.